README.md
# [AutoSys][1]

[![netlify badge](https://api.netlify.com/api/v1/badges/416b8ca3-82db-470f-9adf-a6d06264ca75/deploy-status)][link_netlify] [![Build Status](https://travis-ci.com/skeptycal/autosys.svg?branch=master)][link_travis]

[![Twitter Follow](https://img.shields.io/twitter/follow/skeptycal.svg?style=social)][link_twitter] [![GitHub followers](https://img.shields.io/github/followers/skeptycal.svg?label=GitHub&style=social)][link_github]

---

### **An even more desperate way to make requests for Humans™.**

-   Why does this have to suck so much?
-   Why do my tools make me work for them?
-   Why can't I ask for something I want to know and see it in a way I want to see it?
-   Why haven't my tools changed in 30 years?
-   Why do we need 100 different cryptic patchworks of messy code to get something done?
-   Why do I have to change and contort to accomodate my tools?

## _Why don't my tools work with the way humans think?_

---

### _If computers can do so much, why can't we do any of the things we want to do?_ That is the question that we seek to answer with **_Implore_**.

---

[![test coverage](https://img.shields.io/badge/test_coverage-100%25-6600CC.svg?logo=Coveralls&color=3F5767)](https://coveralls.io) ![CII Best Practices](https://bestpractices.coreinfrastructure.org/projects/3454/badge)

### **Integrates easily in a simple, human way with:**

-   Google Calendars, Contacts, Drive, Docs, Sheets
-   Dictionaries, References, and language tools
-   Analyze news for fake stories and spoofs
-   Professional Journal searches
-   Linux/macOS Command Line
-   simple image edits and pdfs
-   and nearly any public API
-   Latex / BibTex documents
-   macOS Text Messaging
-   Common web browsers
-   Microsoft Excel, Word
-   Reference formatting
-   Text Messaging
-   NLP Processes
-   Apple iTunes
-   Blog Posts
-   Youtube
-   Netflix

_Tell your computer what to do!"_

> Copyright © 2018-2020 [Michael Treanor](https:/skeptycal.github.com) | [MIT License](https://opensource.org/licenses/MIT) - enjoy ...

[![License](https://img.shields.io/badge/License-MIT-darkblue)](https://skeptycal.mit-license.org/1976/) [![macOS Version](https://img.shields.io/badge/macOS-10.15%20Catalina-orange?logo=apple)](https://www.apple.com) [![code style: prettier](https://img.shields.io/badge/code_style-prettier-ff69b4.svg?logo=prettier)](https://github.com/prettier/prettier)

---

## Take a look at Implore

### Features include:

-   **Send text messages** based on results and presets
-   Figure out the annoying **Netflix** recommendations
-   **Watch** for cookies and other local security issues
-   **Research** topics based on logic and language
-   **NLP analysis** of web pages or sets of pages
-   Test **API** functionality and build an interface
-   Perform spoof analysis and **fact checking**
-   **Block** specific activities of certain sites
-   **Convert** websites to 3d object visuals
-   Timing social media and blog **posts**
-   Interact with underlying **database**
-   Create a **command line** interface
-   **Parse** html and language results
-   Watch **Twitter** for specific topics
-   Create link and image **catalogs**
-   Collect **metadata** and tag info
-   **Google Calendar** automation
-   **Collect** website html code
-   **Timestamp** html versions
-   Direct Request from user
-   Create visual **sitemaps**
-   Calendar **reminders**
-   **CSS** Updating
-   **Scripting**
-   **BibTex**
-   **Time**

_Implore requires Python 3.8+ and works best with **lxml** (fast html) and **ujson** (fast json) installed. Text messaging requires macOS, at least until someone tells me a better way to do it._

[![GitHub Pipenv locked Python version](https://img.shields.io/badge/Python-3.8-yellow?color=3776AB&logo=python&logoColor=yellow)](https://www.python.org/) ![Django v3](https://img.shields.io/badge/Django-v3-%23092E20?logo=django&color=#339933)

---

## Contributing

[![Contributor Covenant](https://img.shields.io/badge/Contributor%20Covenant-v1.4%20adopted-ff69b4.svg)](CODE_OF_CONDUCT.md)

**Please feel free to offer suggestions and changes** (contribution instructions below). I have been coding for many years, but mostly as a side activity ... as a tool to assist me in other endeavors ... so I have not had the 'hard time' invested of constant coding that many of you have.

---

## A solid foundation

> Implore uses _Requests_ and _Beautiful Soup_ to parse html and scrape web data. Website output if mostly _Flask_ based. The majority of other functionality is original code.

[1]: (https://www.github.com/skeptycal/autosys)
[link_netlify]: (https://app.netlify.com/sites/mystifying-keller-ab5658/deploys)
[link_travis]: (https://travis-ci.com/skeptycal/autosys)
[link_twitter]: (https://www.twitter.com/skeptycal)
[link_github]: (https://www.github.com/skeptycal)


CHANGELOG.md
# autosys Changelog

## UNRELEASED

-   Finalize Python 3.9 support
-   Remove support for python < 3.8
-   Begin support for PEP 590, Vectorcall: a fast calling protocol for CPython
-   Add default values
-   Add environment variable
-   Update pre-commit, mypy, and tox environment

## autosys 0.2.0

-   Refactor Setup to include sphinx and package_metadata
-   streamline CI path (add Azure and Github hooks)
-   added twine_setup script file to run twine outside of setup.py

## autosys 0.1.0

-   Don't import autosys from setup.py
-   supporting python 3.8+
-   Dev Release. Not Stable


init.sh
#!/usr/bin/env zsh
# -*- coding: utf-8 -*-
    # shellcheck shell=bash
    # shellcheck source=/dev/null
    # shellcheck disable=SC1090,SC2128,SC2178

# setup a github repo and pypi project in the current folder

# it is assumed that the following are installed:
#   python3, git, hub, zsh

#? ------------------------------ Setup Initial Variables
    MAIN="[38;5;229m"
    ATTN="[38;5;178m"
    RESET="[0m"

    # debug_var = echo name and value of variable for debugging
    dv() { echo " ${ATTN}${1}${MAIN} = ${(P)1}${RESET}"; }

    BASH_SOURCE=$(realpath $0)
    REPO_PATH="${BASH_SOURCE%/*}"
    REPO_NAME="${REPO_PATH##*/}"
    TEMPLATE_PATH="$HOME/Documents/coding/template"
    VERSION='0.0.1'
    HOMEPAGE="https://skeptycal.github.io/$REPO_NAME"


    dv BASH_SOURCE
    dv REPO_NAME
    dv REPO_PATH


    exit
#? ------------------------------ Get Repo Info
    vared -p "NAME for python project repo: " -c REPO_NAME
    vared -p 'VERSIOIN for Enter Starting VERSION: ' -c VERSION


    echo "Setup for repo '$REPO_NAME' version ${VERSION}."

    HOMEPAGE="https://skeptycal.github.io/$REPO_NAME"
    DESCRIPTION='System utilities for Python on macOS.'

#? ------------------------------ Setup Repo

python3 -m venv venv
source venv/bin/activate
poetry add -U 
python3 -m pip install -U pip wheel setuptools
mkdir $REPO_NAME
touch $REPO_NAME/__init__.py
mkdir 'testing'

git init
hub create -d $DESCRIPTION -h $HOMEPAGE $REPO_NAME
cp $HOME/
git add .


# ideas ...
# git clone -q https://github.com/skeptycal/autosys.git C:\projects\autosys
# git fetch -q origin +refs/pull/7/merge:
# git checkout -qf FETCH_HEAD
# Running Install scripts
# pip install tox


repo.sh
#!/usr/bin/env zsh
# -*- coding: utf-8 -*-
    # shellcheck shell=bash
    # shellcheck source=/dev/null
    # shellcheck disable=SC1090,SC2128,SC2178

# setup a github repo and pypi project in the current folder

# it is assumed that the following are installed:
#   python3, git, hub, zsh

#? ------------------------------ Setup Initial Variables
    MAIN="[38;5;229m"
    ATTN="[38;5;178m"
    RESET="[0m"

    # debug_var = echo name and value of variable for debugging
    dv() { echo " ${ATTN}${1}${MAIN} = ${(P)1}${RESET}"; }

    BASH_SOURCE=$(realpath $0)
    # SCRIPT_NAME="${BASH_SOURCE##*/}"
    REPO_PATH="${BASH_SOURCE%/*}"
    REPO_NAME="${REPO_PATH##*/}"
    TEMPLATE_PATH="$HOME/Documents/coding/template"
    VERSION='0.0.1'
    HOMEPAGE='https://skeptycal.github.io/$REPO_NAME'


    dv BASH_SOURCE
    dv REPO_NAME
    dv REPO_PATH


    exit
#? ------------------------------ Get Repo Info
    vared -p "Name for python project repo: " -c REPO_NAME
    vared -p 'Enter Starting VERSION: ' -c VERSION


    echo "Setup for repo '$REPO_NAME' version ${VERSION}."

    HOMEPAGE='https://skeptycal.github.io/$REPO_NAME'
    DESCRIPTION='System utilities for Python on macOS.'

#? ------------------------------ Setup Repo
rm -rf $
python3 -m venv venv
source venv/bin/activate
python3 -m pip install -U pip wheel setuptools

git init
hub create -d $DESCRIPTION -h $HOMEPAGE $REPO_NAME
cp $HOME/
git add .


readme.py
from pathlib import Path

from typing import List

try:
    DEFAULT_ENCODING
except:
    from locale import getpreferredencoding
    DEFAULT_ENCODING: str = getpreferredencoding(do_setlocale=True) or "utf-8"


def readme(file_name: str = "readme.md", path_name=__file__) -> (str):
    """ Returns the text of the README file

        The default file is `README.md` and is *NOT* case sensitive. (e.g. `README` is the same as `readme`)
        Can load *any* text file, but the default search path is setup for readme files

        ```
        Search path = ["readme.md", "readme.rst", "readme", "readme.txt"]
        ```

        Example:

        ```
        long_description=readme()
        ```
        """

    search_list: List[str] = ["readme.md",
                              "readme.rst", "readme", "readme.txt"]
    if file_name not in search_list:
        search_list.insert(0, file_name)
    found: bool = False
    for searchfile in search_list:
        for parent in Path(path_name).resolve().parents:
            find_path = Path(parent / searchfile)
            if find_path.exists():
                found = True
                break
        if found:
            break
    if found:
        try:
            with open(find_path, mode="rt", encoding=DEFAULT_ENCODING) as fd:
                return fd.read()
        except IOError as e:
            raise IOError(f"Cannot read from the 'readme' file '{find_path}'")
    else:
        raise FileNotFoundError(
            f"Cannot find project 'readme' file in project tree. Search list = {search_list}"
        )


package_metadata.py
#!/usr/bin/env python3
# -*- coding: utf-8 -*-
""" AutoSys Package Metadata
    ---
    Part of the [AutoSys][1] package

    Copyright (c) 2018 [Michael Treanor][2]

    AutoSys is licensed under the [MIT License][3]

    [1]: https://www.github.com/skeptycal/autosys
    [2]: https://www.twitter.com/skeptycal
    [3]: https://opensource.org/licenses/MIT
    """

from pathlib import Path

from pip_safe import pip_safe_name
from readme import readme

from typing import Dict, List, Tuple

# the default version number is '0.0.1'
__version__: str = '0.0.1'

# put a name here to ignore the default ...
module_name: str = ""

# the default package name is the name of the parent folder ...
if not module_name:
    module_name = Path(__file__).resolve().parent.name

# make the name safe for Pypi.org upload
NAME: str = pip_safe_name(module_name)

VERSION: str = __version__
VERSION_INFO: Tuple[int] = VERSION.split(".")
DESCRIPTION: str = "System utilities for Python on macOS."
REQUIRES_PYTHON: str = ">=3.8.0"
# PACKAGE_DIR: Dict = {f'{NAME}'}
PACKAGE_EXCLUDE: List[str] = ['*test*', '*bak*']
LICENSE: str = "MIT"
LONG_DESCRIPTION: str = readme()
LONG_DESCRIPTION_CONTENT_TYPE: str = "text/markdown"
# LONG_DESCRIPTION_CONTENT_TYPE="text/x-rst",
AUTHOR: str = "Michael Treanor"
AUTHOR_EMAIL: str = "skeptycal@gmail.com"
MAINTAINER: str = ""
MAINTAINER_EMAIL: str = ""
URL: str = f"https://skeptycal.github.io/{NAME}/"
DOWNLOAD_URL: str = f"https://github.com/skeptycal/{NAME}/archive/{VERSION}.tar.gz"
ZIP_SAFE: bool = False
INCLUDE_PACKAGE_DATA: bool = True
# What packages are required for this module to be executed?
REQUIRED: List[str] = [
    "aiocontextvars>=0.2.0 ; python_version<'3.7'",
    "colorama>=0.3.4 ; sys_platform=='win32'",
    "win32-setctime>=1.0.0 ; sys_platform=='win32'",
]

# What packages are optional?
EXTRAS: Dict = {}

PACKAGE_DATA: Dict = {
    # If any package contains these files, include them:
    "": [
        "*.txt",
        "*.rst",
        "*.md",
        "*.ini",
        "*.png",
        "*.jpg",
        "__init__.pyi",
        "py.typed",
    ]
}

PROJECT_URLS: Dict = {
    "Website": f"https://skeptycal.github.io/{NAME}/",
    "Documentation": f"https://skeptycal.github.io/{NAME}/docs",
    "Source Code": f"https://www.github.com/skeptycal/{NAME}/",
    "Changelog":
    f"https://github.com/skeptycal/{NAME}/blob/master/CHANGELOG.md",
}

KEYWORDS: List = [
    "application",
    "macOS",
    "dev",
    "devops",
    "cache",
    "utilities",
    "cli",
    "python",
    "cython",
    "text",
    "console",
    "log",
    "debug",
    "test",
    "testing",
    "logging",
    "logger",
]
CLASSIFIERS: List = [
    "Development Status :: 4 - Beta",
    "License :: OSI Approved :: MIT License",
    "Environment :: Console",
    "Environment :: MacOS X",
    "Intended Audience :: Developers",
    "Natural Language :: English",
    "Operating System :: MacOS",
    "Operating System :: OS Independent",
    "Programming Language :: Cython",
    "Programming Language :: Python",
    "Programming Language :: Python :: 3 :: Only",
    "Programming Language :: Python :: 3",
    # These are the Python versions tested; it may work on others
    "Programming Language :: Python :: 3.8",
    "Programming Language :: Python :: 3.9",
    "Programming Language :: Python :: Implementation :: CPython",
    "Programming Language :: Python :: Implementation :: PyPy",
    "Topic :: Software Development :: Libraries :: Python Modules",
    "Topic :: Software Development :: Testing",
    "Topic :: Utilities",
]
